pennylane/compiler/compiler.py

```diff
@@ -0,0 +1,204 @@
+
+    # Refresh the list of compilers
+    AvailableCompilers.names_entrypoints = defaultdict(dict)
+
+    # Iterator packages entry-points with the 'pennylane.compilers' group name
+    for entry in pkg_resources.iter_entry_points("pennylane.compilers"):
+        # Only need name of the parent module
+        module_name = entry.module_name.split(".")[0]
+        AvailableCompilers.names_entrypoints[module_name][entry.name] = entry
+
+    # Check whether available compilers follow the entry_point interface
+    # by validating that all entry points (qjit, context, and ops) are defined.
+    for _, eps_dict in AvailableCompilers.names_entrypoints.items():
+        ep_interface = AvailableCompilers.entrypoints_interface
+        if any(ep not in eps_dict.keys() for ep in ep_interface):
+            raise KeyError(f"expected {ep_interface}, but recieved {eps_dict}")
+
+
+# Scan installed compiler packages
+# and update AvailableCompilers
+_refresh_compilers()
+
+
+def _reload_compilers():
+    """Reload and scan installed PennyLane compiler packages to refresh the
+    compilers names and entry points.
+    """
+
+    reload(pkg_resources)
+    _refresh_compilers()
+
+
+def available_compilers() -> List[str]:
+    """Loads and returns a list of available compilers that are
+    installed and compatible with the :func:`~.qjit` decorator.
+
+    **Example**
+
+    This method returns the name of installed compiler packages supported in
+    PennyLane. For example, after installing the
+    `Catalyst <https://github.com/pennylaneai/catalyst>`__
+    compiler, this will now appear as an available compiler:
+
+    >>> qml.compiler.available_compilers()
+    ['catalyst']
+    """
+
+    # Reload installed packages and updates
+    # the class variable names_entrypoints
+    _reload_compilers()
+
+    return list(AvailableCompilers.names_entrypoints.keys())
+
+
+def available(compiler="catalyst") -> bool:
+    """Check the availability of the given compiler package.
+
+    Args:
+        compiler (str): name of the compiler package (default value is ``catalyst``)
+
+    Return:
+        bool: ``True`` if the compiler package is installed on the system
+
+    **Example**
+
+    Before installing the ``pennylane-catalyst`` package:
+
+    >>> qml.compiler.available("catalyst")
+    False
+
+    After installing the ``pennylane-catalyst`` package:
+
+    >>> qml.compiler.available("catalyst")
+    True
+    """
+
+    # It only refreshes the compilers names and entry points if the name
+    # is not already stored. This reduces the number of re-importing
+    # ``pkg_resources`` as it can be a very slow operation on systems
+    # with a large number of installed packages.
+
+    if compiler not in AvailableCompilers.names_entrypoints:
+        # Reload installed packages and updates
+        # the class variable names_entrypoints
+        _reload_compilers()
+
+    return compiler in AvailableCompilers.names_entrypoints
+
+
+def active_compiler() -> str:
```

Comment:
```suggestion
def active_compiler() -> str:
```
should the type hint include `None` as well?

Reply:
Tbh I don't know what the best practice is here in Python but I saw it in some code examples. Let me look into PL to see if there're any favorable style for these cases.